storage/pebble/registers_pruner.go

```diff
@@ -0,0 +1,283 @@
+	}
+
+	if firstHeight > latestHeight {
+		return errors.New("the latest height must be greater than the first height")
+	}
+
+	if latestHeight-firstHeight <= p.pruneInterval {
+		return nil
+	}
+
+	pruneHeight := latestHeight - p.pruneThreshold
+	p.logger.Info().Uint64("prune_height", pruneHeight).Msg("pruning storage")
+
+	if err = p.pruneUpToHeight(ctx, p.db, pruneHeight); err != nil {
+		return fmt.Errorf("failed to prune: %w", err)
+	}
+
+	return nil
+}
+
+// pruneUpToHeight prunes all entries in the database with heights less than or equal
+// to the specified pruneHeight. For each register prefix, it keeps the earliest entry
+// that has a height less than or equal to pruneHeight, and deletes all other entries
+// with lower heights.
+//
+// This function iterates over the database keys, identifies keys to delete in batches,
+// and uses the batchDelete function to remove them efficiently.
+//
+// Parameters:
+//   - ctx: The context for managing the pruning throttle delay operation.
+//   - pruneHeight: The maximum height of entries to prune.
+//
+// No errors are expected during normal operations.
+func (p *RegisterPruner) pruneUpToHeight(ctx context.Context, r pebble.Reader, pruneHeight uint64) error {
+	start := time.Now()
+
+	// first, update firstHeight in the db
+	// this ensures that if the node crashes during pruning, there will still be a consistent
+	// view when the node starts up. Subsequent prunes will remove any leftover data.
+	if err := p.updateFirstStoredHeight(pruneHeight); err != nil {
+		return fmt.Errorf("failed to update first height for register storage: %w", err)
+	}
+
+	dbPruner := RegisterPrunerRun(p.db, p.logger, p.pruneThrottleDelay, pruneHeight)
+
+	prefix := []byte{codeRegister}
+	it, err := r.NewIter(&pebble.IterOptions{
+		LowerBound: prefix,
+		UpperBound: []byte{codeRegister + 1},
+	})
+	if err != nil {
+		return fmt.Errorf("cannot create iterator: %w", err)
+	}
+
+	defer func() {
+		if cerr := it.Close(); cerr != nil {
+			p.logger.Err(cerr).Msg("error while closing the iterator")
+		}
+	}()
+
+	var batchKeysToRemove [][]byte
+
+	for it.SeekGE(prefix); it.Valid(); it.Next() {
+		key := it.Key()
+
+		canPruneKey, err := dbPruner.CanPruneKey(key)
+		if err != nil {
+			return fmt.Errorf("cannot check key %v for pruning, reason: %w", key, err)
+		}
+
+		if !canPruneKey {
+			continue
+		}
+
+		// Create a copy of the key to avoid memory issues
+		batchKeysToRemove = append(batchKeysToRemove, bytes.Clone(key))
+
+		if len(batchKeysToRemove) >= deleteItemsPerBatch {
+			// Perform batch delete
+			if err := dbPruner.BatchDelete(ctx, batchKeysToRemove); err != nil {
+				return err
+			}
+
+			// Throttle to prevent excessive system load
+			select {
+			case <-ctx.Done():
+			case <-time.After(p.pruneThrottleDelay):
+			}
+
+			// Reset batchKeysToRemove to empty slice while retaining capacity
+			batchKeysToRemove = batchKeysToRemove[:0]
```

Comment:
I thought about this as well. it turns our `clear()` called on a slice just sets all of the elements in the array to nil rather than changing the size of slice.